[Swift in WebKit] Add initial support for Swift-Cxx interop in WebKit

rr-codes opened the pull request:
#### 7aa2ee380a72ee199901e6eaeec3006c78d13be8
<pre>
[Swift in WebKit] Add initial support for Swift-Cxx interop in WebKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=296393">https://bugs.webkit.org/show_bug.cgi?id=296393</a>
<a href="https://rdar.apple.com/156525628">rdar://156525628</a>

Reviewed by Abrar Rahman Protyasha.

Set `SWIFT_OBJC_INTEROP_MODE` to `objcxx` in WebKit so that Swift code in WebKit can access C++.

Specifically, any header file that is part of a clang module (such as WebKit_Internal
and WebKit_Internal_Cxx) can be accessed from Swift. However, each of those files need
to be compatible with modularization, and must include all headers they use (and cannot
for example depend on transitive imports).

Currently, enable this only for macOS due to some module issues that affect iOS and Catalyst.
Once these are resolved, the switch can be flipped for those platforms.

* Source/WebCore/PAL/pal/spi/cocoa/WritingToolsSPI.h:
* Source/WebCore/PAL/pal/spi/cocoa/WritingToolsUISPI.h:

Due to an issue in the header files of WritingTools, an error happens if they are
attempted to be accessed as modularized headers. To workaround this, the imports
themselves are removed in favor of just forward declaring the symbols, just like the
open source build.

In doing so, this reveals some minor differences between the forward declarations and the actual header contents, so those are fixed as well.

* Source/WebCore/platform/LogClient.h:
* Source/WebKit/Configurations/WebKit.xcconfig:

Set `SWIFT_OBJC_INTEROP_MODE = objcxx;`, which overrides the `SWIFT_OBJC_INTEROP_MODE =
objc` value in Base.xcconfig so that WebKit can use the interop without WebKitSwift or
others needing it.

* Source/WebKit/Modules/Internal/WebKitInternalCxx.h: Added.
* Source/WebKit/Modules/Internal/module.modulemap:

Add a new module for non-ObjC C++ headers, WebKit_Internal_Cxx.

* Source/WebKit/Platform/IPC/Connection.h:
* Source/WebKit/Platform/IPC/ConnectionHandle.h:
* Source/WebKit/Platform/IPC/Decoder.h:
* Source/WebKit/Platform/IPC/MessageReceiveQueueMap.h:
* Source/WebKit/Platform/IPC/StreamConnectionBuffer.h:
* Source/WebKit/Platform/LogClient.h:
* Source/WebKit/Scripts/webkit/messages.py:
(generate_message_names_header):

These headers are dependencies of some file in the WebKit_Internal or
WebKit_Internal_Cxx clang modules. As a result, add any imports that were previously
absent due to having been imported transitively.

* Source/WebKit/UIProcess/API/Cocoa/WKContextMenuElementInfoAdapter.swift:
* Source/WebKit/UIProcess/API/Cocoa/WebKitSwiftOverlay.swift:
(WKWebView.find(_:configuration:)):
(WKWebView.evaluateJavaScript(_:in:contentWorld:)):
* Source/WebKit/UIProcess/API/Swift/URLSchemeHandler.swift:
* Source/WebKit/UIProcess/API/Swift/WebPage+BackForwardList.swift:
* Source/WebKit/UIProcess/API/Swift/WebPage+Configuration.swift:
(Configuration.applicationNameForUserAgent):
* Source/WebKit/UIProcess/API/Swift/WebPage+DialogPresenting.swift:
(DialogPresenting.handleJavaScriptAlert(_:initiatedBy:)):
(DialogPresenting.handleJavaScriptConfirm(_:initiatedBy:)):
(WebPage.handleJavaScriptAlert(_:initiatedBy:)):
(WebPage.handleJavaScriptConfirm(_:initiatedBy:)):
* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(url):
(title):
(customUserAgent):
(load(_:)):
(load(_:mimeType:characterEncoding:baseURL:)):
(load(_:baseURL:)):
(load(simulatedRequest:responseHTML:)):
* Source/WebKit/UIProcess/Cocoa/Foundation+Extras.swift:

Now that Swift can access C++, there are some naming conflicts with `Swift.String` with
`wtf/String`, and `Foundation.URL` with `wtf/URL`. To resolve this, a typealias is
defined so that the Swift versions take precedence in Swift. For API declarations, the
fully qualified symbol name must be used because the typealias itself is not public and
should not be.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/297841@main">https://commits.webkit.org/297841@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fb5f1a274f2ced0eb11ccfe75c06627831f85287

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113103 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32838 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23316 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119310 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63702 "Built successfully") | ✅ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115065 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33490 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41401 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/86086 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/37589 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116050 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26745 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101744 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66404 "Passed tests") | | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/112535 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26021 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19874 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63065 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96132 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19948 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122528 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40181 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29971 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/94931 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40565 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97962 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94672 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; run-api-tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24156 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39821 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/17626 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/36306 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40067 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45566 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Checked out pull request; Found 70413 issues") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39708 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43041 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41445 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/UIProcess/Cocoa/Foundation+Extras.swift:
```diff
@@ -21,10 +21,13 @@
 
 import Foundation
 
+typealias String = Swift.String
+typealias URL = Foundation.URL
```
So to access `WTFString.h`, we would want to use `WTF::String`?

rr-codes replied:
to access it from Swift (which you probably wouldn't need to do at all) you would do `wtf.String` (wtf being the module name)

aprotyas replied:
okay! Just curious how the syntax looks like

aprotyas commented on Source/WebKit/Configurations/WebKit.xcconfig:
```diff
@@ -355,6 +355,12 @@ WK_MAC_PUBLIC_IOS_PRIVATE_HEADERS_DIR_cocoatouch = $(PRIVATE_HEADERS_FOLDER_PATH
+
+// FIXME: (rdar://149474443) Swift/Cxx interop does not respect CLANG_CXX_LANGUAGE_STANDARD = "c++2b";
+// FIXME: (rdar://103177280) "-no-verify-emitted-module-interface;" is needed to workaround a SwiftVerifyEmittedModuleInterface bug
+OTHER_SWIFT_FLAGS = $(inherited) -Xcc -std=c++2b -no-verify-emitted-module-interface;
```
Would we not want something like this?
```suggestion
OTHER_SWIFT_FLAGS = $(inherited) -Xcc -std=$(CLANG_CXX_LANGUAGE_STANDARD) -no-verify-emitted-module-interface;
```

rr-codes replied:
oh, ya maybe that would work! I'll try it

aprotyas reviewed (approved):
Looks good, couple of questions

adetaylor commented on Source/WebKit/UIProcess/Cocoa/Foundation+Extras.swift:
```diff
@@ -21,10 +21,13 @@
 
 import Foundation
 
+typealias String = Swift.String
```
Why have you done this here but in lots of the other files you've full qualified each use of `String`? (I think this `typealias` approach is probably better, but either way, consistency seems good?)

rr-codes replied:
I explained this a bit in my commit message:

> Now that Swift can access C++, there are some naming conflicts with `Swift.String` with
`wtf/String`, and `Foundation.URL` with `wtf/URL`. To resolve this, a typealias is
defined so that the Swift versions take precedence in Swift. For API declarations, the
fully qualified symbol name must be used because the typealias itself is not public and
should not be.

so basically, the vast majority of the time you can just use String and URL, but for API symbols specifically you will need to fully qualify the name. I believe this is an acceptable compromise since the amount of non-API code is vastly larger than API code.

adetaylor replied:
Ah thanks. Makes sense.

rr-codes committed:
[Swift in WebKit] Add initial support for Swift-Cxx interop in WebKit
https://bugs.webkit.org/show_bug.cgi?id=296393
rdar://156525628

Reviewed by Abrar Rahman Protyasha.

Set `SWIFT_OBJC_INTEROP_MODE` to `objcxx` in WebKit so that Swift code in WebKit can access C++.

Specifically, any header file that is part of a clang module (such as WebKit_Internal
and WebKit_Internal_Cxx) can be accessed from Swift. However, each of those files need
to be compatible with modularization, and must include all headers they use (and cannot
for example depend on transitive imports).

Currently, enable this only for macOS due to some module issues that affect iOS and Catalyst.
Once these are resolved, the switch can be flipped for those platforms.

* Source/WebCore/PAL/pal/spi/cocoa/WritingToolsSPI.h:
* Source/WebCore/PAL/pal/spi/cocoa/WritingToolsUISPI.h:

Due to an issue in the header files of WritingTools, an error happens if they are
attempted to be accessed as modularized headers. To workaround this, the imports
themselves are removed in favor of just forward declaring the symbols, just like the
open source build.

In doing so, this reveals some minor differences between the forward declarations and the actual header contents, so those are fixed as well.

* Source/WebCore/platform/LogClient.h:
* Source/WebKit/Configurations/WebKit.xcconfig:

Set `SWIFT_OBJC_INTEROP_MODE = objcxx;`, which overrides the `SWIFT_OBJC_INTEROP_MODE =
objc` value in Base.xcconfig so that WebKit can use the interop without WebKitSwift or
others needing it.

* Source/WebKit/Modules/Internal/WebKitInternalCxx.h: Added.
* Source/WebKit/Modules/Internal/module.modulemap:

Add a new module for non-ObjC C++ headers, WebKit_Internal_Cxx.

* Source/WebKit/Platform/IPC/Connection.h:
* Source/WebKit/Platform/IPC/ConnectionHandle.h:
* Source/WebKit/Platform/IPC/Decoder.h:
* Source/WebKit/Platform/IPC/MessageReceiveQueueMap.h:
* Source/WebKit/Platform/IPC/StreamConnectionBuffer.h:
* Source/WebKit/Platform/LogClient.h:
* Source/WebKit/Scripts/webkit/messages.py:
(generate_message_names_header):

These headers are dependencies of some file in the WebKit_Internal or
WebKit_Internal_Cxx clang modules. As a result, add any imports that were previously
absent due to having been imported transitively.

* Source/WebKit/UIProcess/API/Cocoa/WKContextMenuElementInfoAdapter.swift:
* Source/WebKit/UIProcess/API/Cocoa/WebKitSwiftOverlay.swift:
(WKWebView.find(_:configuration:)):
(WKWebView.evaluateJavaScript(_:in:contentWorld:)):
* Source/WebKit/UIProcess/API/Swift/URLSchemeHandler.swift:
* Source/WebKit/UIProcess/API/Swift/WebPage+BackForwardList.swift:
* Source/WebKit/UIProcess/API/Swift/WebPage+Configuration.swift:
(Configuration.applicationNameForUserAgent):
* Source/WebKit/UIProcess/API/Swift/WebPage+DialogPresenting.swift:
(DialogPresenting.handleJavaScriptAlert(_:initiatedBy:)):
(DialogPresenting.handleJavaScriptConfirm(_:initiatedBy:)):
(WebPage.handleJavaScriptAlert(_:initiatedBy:)):
(WebPage.handleJavaScriptConfirm(_:initiatedBy:)):
* Source/WebKit/UIProcess/API/Swift/WebPage.swift:
(url):
(title):
(customUserAgent):
(load(_:)):
(load(_:mimeType:characterEncoding:baseURL:)):
(load(_:baseURL:)):
(load(simulatedRequest:responseHTML:)):
* Source/WebKit/UIProcess/Cocoa/Foundation+Extras.swift:

Now that Swift can access C++, there are some naming conflicts with `Swift.String` with
`wtf/String`, and `Foundation.URL` with `wtf/URL`. To resolve this, a typealias is
defined so that the Swift versions take precedence in Swift. For API declarations, the
fully qualified symbol name must be used because the typealias itself is not public and
should not be.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/297841@main
